runtime/src/qio/qio_formatted.c

```diff
@@ -361,6 +361,91 @@ qioerr _append_char(char* restrict * restrict buf, size_t* restrict buf_len, siz
   return 0;
 }
 
+qioerr read_entire_file(const int threadsafe,qio_channel_t* restrict ch, const char* restrict* restrict out, int64_t* restrict len_out, ssize_t maxlen)
+{  
+  qioerr err,err_reading_file=0;
+  int64_t file_length,chunkSize=1024;
+  char* ret=NULL;
+  ssize_t amt=0;
+  err_t errcode;
+  ssize_t total_amt_read=0;
+  bool err_flag;
+  if( threadsafe ) 
+  {
+    err = qio_lock(&ch->lock);
+    if( err ) return err;
+  }
+  // obtain file length
+  err=qio_file_length(ch->file,&file_length);   
+  // no error in obtaining file length
+  if(!err)                                     
+  {
+    //allocate initial buffer
+    ret=qio_malloc(file_length+1);              
+    ret[0]='\0';
+    //file_length+1 instead of file_length -->I try to read one more character than 
+   //the prescribed file length . That way, if we have already reached the file end,
+   //reading 1 more character will generate EOF. But, if the file length has increased
+   //while reading, we have just read one more character.
+    err=qio_channel_read(false, ch, ret,file_length+1, &amt);    
+    // Here,checking whether file end is reached.
+    if(qio_err_to_int(err)==EEOF)                                
+    {
+       err=0;
+    }
+    else
+    {
+       err=QIO_EEOF;
+    }
+    total_amt_read=amt;
+  }
+  // error might be due to being unable to fetch file length or there is 
+  //something left to read.
+  if(err)                                                        
+  {    
+    // add some more space to the buffer to read the remaining data.
+         
+    while(true)
+    {
+      ret=qio_realloc(ret,total_amt_read+chunkSize+1);       
+      // we will read the remaining data in units of 1024 each time .
+      //calculate the amt that we have read until now
+      // break out of loop is data is exhausted finally
+      err_reading_file = qio_channel_read(false, ch,&ret[total_amt_read],chunkSize, &amt);   
+      total_amt_read+=amt;                                                                
+      if (err_reading_file == QIO_EEOF) break;                                            
+      
```

Comment:
This loop used to double chunkSize on each iteration, you could do that here at the end simply with `chunkSize *= 2`. That prevents the reallocation process from becoming O(n*n).

Reply:
@mppf What I am basically doing is reading 1024 characters each time. So,why would the relocation be O(n*n)?

Reply:
the call to realloc is O(size of memory being reallocated). You are calling it in a loop that is O(size of memory / 1024). So it is O(size * size).